compactor2/src/components/combos/throttle_partition.rs

```diff
@@ -0,0 +1,507 @@
+///          +...........................................+
+/// ```
+///
+/// | Step |  Name                 | Type                                                              | Description |
+/// | ---- | --------------------- | ----------------------------------------------------------------- | ----------- |
+/// | 1    | **Actual source**     | `inner_source`/`T1`/[`PartitionsSource`], wrapped                 | This is the actual source. |
+/// | 2    | **Throttling source** | [`ThrottlePartitionsSourceWrapper`], wraps `inner_source`/`T1`    | Throttles partitions that do not receive any commits |
+/// | 3    | **Critical section**  | --                           | The actual partition processing    |
+/// | 4    | **Throttle commit**   | [`ThrottleCommitWrapper`], wraps `inner_commit`/`T2`              | Observes commits. |
+/// | 5    | **Actual commit**     | `inner_commit`/`T2`/[`Commit`] | The actual commit implementation |
+/// | 6    | **Throttle sink**     | [`ThrottlePartitionDoneSinkWrapper`], wraps `inner_sink`/`T3`     | Observes incoming IDs enables throttled if step (4) did not observe any commits. |
+/// | 7    | **Actual sink**       | `inner_sink`/`T3`/[`PartitionDoneSink`], wrapped                  | The actual sink.  Directly receives all partitions filtered out at step 2. |
+///
+/// Note that partitions filtered out by [`ThrottlePartitionsSourceWrapper`] will directly be forwarded to `inner_sink`. No
+/// partition is ever lost. This means that `inner_source` and `inner_sink` can perform proper accounting. The
+/// concurrency of this bypass can be controlled via `bypass_concurrency`.
+///
+/// This setup relies on a fact that it does not process duplicate [`PartitionId`]. You may use
+/// [`unique_partitions`](crate::components::combos::unique_partitions::unique_partitions) to achieve that.
+pub fn throttle_partition<T1, T2, T3>(
+    source: T1,
+    commit: T2,
+    sink: T3,
+    time_provider: Arc<dyn TimeProvider>,
+    throttle_duration: Duration,
+    bypass_concurrency: usize,
+) -> (
+    ThrottlePartitionsSourceWrapper<T1, T3>,
+    ThrottleCommitWrapper<T2>,
+    ThrottlePartitionDoneSinkWrapper<T3>,
+)
+where
+    T1: PartitionsSource,
+    T2: Commit,
+    T3: PartitionDoneSink,
+{
+    let state = SharedState::default();
+    let inner_sink = Arc::new(sink);
+    let source = ThrottlePartitionsSourceWrapper {
+        inner_source: source,
+        inner_sink: Arc::clone(&inner_sink),
+        state: Arc::clone(&state),
+        time_provider: Arc::clone(&time_provider),
+        sink_concurrency: bypass_concurrency,
+    };
+    let commit = ThrottleCommitWrapper {
+        inner: commit,
+        state: Arc::clone(&state),
+    };
+    let sink = ThrottlePartitionDoneSinkWrapper {
+        inner: inner_sink,
+        state,
+        time_provider,
+        throttle_duration,
+    };
+    (source, commit, sink)
+}
+
+#[derive(Debug, Default)]
+struct State {
+    in_flight: HashMap<PartitionId, bool>,
```

Comment:
Have I understood this right?

```suggestion
    // Value is "true" while compaction task is in-flight, and "false" once complete.
    // 
    // Completed compaction tasks are removed from the map each time the source fetch()
    // is called.
    in_flight: HashMap<PartitionId, bool>,
```